### PR TITLE
Store environment variables in .env

### PR DESCRIPTION
**Pull request by piguyinthesky**
Instead of having to run `export MY_PLATFORM=local`, etc. each time you start up a new shell, a much cleaner solution is storing the environment variables inside a `.env` file in the root directory.

The `dotenv` package automatically loads the key-value pairs stored in `.env` into your `process.env` so that they are accessible as environment variables.

### CODE REVIEW

**Comment by vck3000** [comment]
Interesting suggestion, I like it! Is it possible to create a default .env file if the file doesn't exist? 

Also, are many of the additional lines in `.gitignore` actually used in this project? Or did you copy paste a bunch of standard files that should be ignored? (I'm ok with this)

**Comment by piguyinthesky** [comment]
Yeah, I copied the template from node.js -- I just like the way it looks, tbh it's a bit of an unnecessary change. I just wasn't sure why `passport-local-mongoose` was the only node module being excluded.

I have a "`.env.example`" put in the repository. Typically the `.env` file contains confidential information, so you don't want to commit it to your repository, but there's no passwords and stuff in this one so I think it should be fine.

Yeah, one problem with my setup is that I couldn't figure out how to make permanent local environment variables; having a `.env` file works for all systems and I think it's clearer overall.

**Comment by vck3000** [comment]
I had to specifically edit `passport-local-mongoose` to make usernames case insensitive, so that's why that is left in the source code here (and `.gitignore`d).

The env.example is good. I was thinking more of: 
When app.js is run, it will try to find the `.env` file. If it doesn't exist, create a file with the default values.

**Comment by vck3000** [comment]
I just realised I hadn't considered this: how will this affect the actual deployed server's environment variables?

**Comment by piguyinthesky** [comment]
Oh, you mentioned you were hosting it using heroku? You can [set heroku config vars](https://devcenter.heroku.com/articles/config-vars).

Also, I think you can make those default values in the code itself, i.e. `process.env.myPlatform || "local"`, or you could directly set it at the start of `app.js`:
```
if (typeof process.env.MY_PLATFORM === "undefined") {
    process.env.MY_PLATFORM = "local";
}
```
That's a little bit tacky, though, since usually you aren't supposed to assign values onto process.env in your code, but it could be a quick fix.

**Comment by piguyinthesky** [comment]
Also, I don't think editing the library directly is a great idea... Did you know you can do the same thing by specifying "`options.usernameLowerCase`" in your options for `passportLocalMongoose?` I'll take another look into it, but I think this is an issue for another pull request. I'll open one up as I figure it out.

**Comment by piguyinthesky** [comment]
Wait, doesn't each user have a "usernameLower" field? Why not just use that in our queries if it's already built in?

File: app.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 //=====================================
 //INITIALISATION
 //=====================================
+require("dotenv").config();
```
Comment (vck3000) [inline app.js]: How would this line affect pre-existing environment variables? If I have a permanent local variable on the deployed server with `MY_PLATFORM=server`, would that get overwritten?

**Comment by vck3000** [comment]
At the moment I am setting the config vars through heroku's online interface and it's been working well. But I suspect they are simply setting permanent environment variables, and I'm not sure how dotenv will affect it. If you could confirm that no conflict would occur, then it'll be all good!

I would like to avoid the tacky solution you've proposed there if possible! 

> Did you know you can do the same thing by specifying "options.usernameLowerCase" in your options for passportLocalMongoose?

I actually don't know how I did not know this. After looking back at my code, it looks like I'm not using that library, but instead using one that could combine passport, mongoose and sockets all together.

I actually hacked it together without really knowing much detail, so if you have a better idea than me, please do let me know :). 

Each user has a usernameLower attribute which I only assigned much after I had deployed the server. It was mainly due to problems with searching up certain users by their case-sensitive usernames. I'm unsure how to make passport look at the lower cased username attribute.

File: app.js (hunk 1 above)
Comment (piguyinthesky) [reply app.js]: Yes, I think so. You would just put `MY_PLATFORM=server` into your `.env` file instead.

**Comment by piguyinthesky** [comment]
Oh ok so the reason you add `.env` to your `.gitignore` is so that it doesn't actually get pushed to the Heroku server. The `require("dotenv").config()` line doesn't do anything on Heroku since all of the environment variables are set by Heroku itself.

**Comment by piguyinthesky** [comment]
Hmm yeah, with different passport libraries, I've found that there are so many different ways you can do things that it's frankly usually a bother... Imo, the fewer libraries the better. We can just query for the user using their `usernameLower` attribute instead. I don't have a ton of experience with authentication in passport, but I'll take a look and see if there's a simpler way to do things.

**Comment by vck3000** [comment]
Just tested this PR, it breaks the username capitalisation...

**Pull request closed without merging**